File: apps/mybookkeeper/backend/app/schemas/properties/utility_offer.py
"""A single electricity offer available at a ZIP code, as ranked for the operator."""
from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel


class UtilityOffer(BaseModel):
    # Power to Choose's own plan id. Not a local row — nothing about an offer is
    # persisted until the operator actually signs up and records the new plan.
    external_plan_id: int
    provider_name: str
    plan_name: str
    term_months: int
    # All three published disclosure points, because the shape across them is
    # what separates a real price from a bill-credit teaser. The client shows
    # all three rather than making the operator trust a single number.
    price_cents_per_kwh_at_500: Decimal
    price_cents_per_kwh_at_1000: Decimal
    price_cents_per_kwh_at_2000: Decimal
    renewable_pct: int | None = None
    # Power to Choose customer rating, 1-5. None when the feed has none on file
    # — which the UI must render as "unrated", never as a zero-star score.
    provider_rating: int | None = None
    # J.D. Power residential satisfaction score, 1-5, where published. Only a
    # minority of REPs carry one, so it is a bonus signal and never a gate.
    jd_power_rating: int | None = None
    # None means the feed stated no dollar figure — distinct from a stated $0.
    cancellation_fee_cents: int | None = None
    cancellation_fee_is_per_remaining_month: bool = False
    # True when the 1000 kWh headline is far below this plan's own price at 500
    # and 2000 kWh. Ranked below honest plans and badged in the UI.
    is_teaser_priced: bool = False
    # Signed cents per year at the reference usage, against the plan currently
    # held for this property. Negative means the offer is worse.
    annual_saving_cents: int | None = None
    # The provider's Electricity Facts Label — the legally binding terms. Always
    # surfaced, because a ranking is a starting point and not a recommendation.
    fact_sheet_url: str | None = None
    enroll_url: str | None = None

File: apps/mybookkeeper/frontend/src/shared/types/utility/utility-offer.ts
/**
 * One electricity offer available at a property's ZIP, as ranked by the API.
 *
 * Nothing here is a local row — an offer becomes a `UtilityPlan` only if the
 * operator signs up and records it through the normal add flow.
 *
 * Mirrors `backend/app/schemas/properties/utility_offer.py`.
 */
export interface UtilityOffer {
  external_plan_id: number;
  provider_name: string;
  plan_name: string;
  term_months: number;
  /** Decimals arrive as strings so no precision is lost in JSON. */
  price_cents_per_kwh_at_500: string;
  price_cents_per_kwh_at_1000: string;
  price_cents_per_kwh_at_2000: string;
  renewable_pct: number | null;
  /** 1-5. `null` means unrated — never render that as zero stars. */
  provider_rating: number | null;
  jd_power_rating: number | null;
  /** `null` means no figure was published — distinct from a stated $0. */
  cancellation_fee_cents: number | null;
  cancellation_fee_is_per_remaining_month: boolean;
  is_teaser_priced: boolean;
  /**
   * Always `null` on a time-of-use offer — the reference usage says nothing
   * about when power is drawn, so there is no honest figure to put here.
   */
  annual_saving_cents: number | null;
  /**
   * The plan prices power differently by hour or day — free nights, free
   * weekends, off-peak discounts. The three disclosure prices are then a
   * blended average, not a rate that will be paid, so these are listed apart
   * from the ranked offers and never compared against a flat plan.
   */
  is_time_of_use: boolean;
  /**
   * The provider's own description of what the plan includes, verbatim and
   * truncated. Only populated for time-of-use offers, where it is the one place
   * the free window is stated in words.
   */
  special_terms: string | null;
  fact_sheet_url: string | null;
  enroll_url: string | null;
}
